File: backend/app/schemas/connectivity.py
"""Pydantic schemas for Fase 18 — Network Connectivity Analysis."""
from __future__ import annotations

from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict


class RouteEntry(BaseModel):
    destination: str
    prefix_len:  int
    next_hop:    str
    interface:   str | None = None
    protocol:    str        = "static"
    distance:    int        = 0
    metric:      int        = 0
    active:      bool       = True


class BgpPeer(BaseModel):
    peer_ip:  str
    asn:      str | None = None
    state:    str        = "unknown"
    uptime:   str | None = None
    prefixes_received: int = 0


class OspfNeighbor(BaseModel):
    neighbor_id: str
    state:       str        = "unknown"
    interface:   str | None = None
    address:     str | None = None


class ConnectivityAnomaly(BaseModel):
    type:        str
    severity:    str
    description: str
    details:     dict[str, Any] | None = None


class PairAnalysisRequest(BaseModel):
    device_b_id: UUID


class ConnectivityAnalysisSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id:            str
    tenant_id:     str | None
    device_id:     str
    mode:          str = "single"
    device_b_id:   str | None = None
    status:        str
    anomaly_count: int = 0
    route_count:   int = 0
    created_at:    str
    completed_at:  str | None = None
    error:         str | None = None


class ConnectivityAnalysisRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id:           str
    tenant_id:    str | None
    device_id:    str
    mode:         str = "single"
    device_b_id:  str | None = None
    status:       str

    # Dispositivo A (ou único em modo single)
    routes:             list[dict] | None = None
    bgp_peers:          list[dict] | None = None
    ospf_neighbors:     list[dict] | None = None
    sdwan_services:     list[dict] | None = None

    # Dispositivo B — apenas mode="pair"
    device_b_routes:          list[dict] | None = None
    device_b_bgp_peers:       list[dict] | None = None
    device_b_ospf_neighbors:  list[dict] | None = None
    device_b_sdwan_services:  list[dict] | None = None

    anomalies:          list[dict] | None = None
    ai_summary:         str | None = None
    ai_recommendations: list[str] | None = None
    error:              str | None = None
    created_at:         str
    completed_at:       str | None = None

File: frontend/src/types/connectivity.ts
export type ConnectivityStatus = "pending" | "running" | "completed" | "failed";
export type ConnectivityMode   = "single" | "pair";

export interface ConnectivityAnomalySeverity {
  type: string;
  severity: "high" | "medium" | "low";
  description: string;
  details?: Record<string, unknown>;
  _scope?: "pair";      // cross-device anomalies
  _device?: string;     // which device originated this anomaly
}

export interface RouteEntry {
  destination: string;
  prefix_len: number;
  next_hop: string;
  interface?: string;
  protocol: string;
  active: boolean;
}

export interface BgpPeer {
  peer_ip: string;
  asn?: string;
  state: string;
  uptime?: string;
  prefixes_received: number;
}

export interface OspfNeighbor {
  neighbor_id: string;
  state: string;
  interface?: string;
  address?: string;
}

export interface SdwanService {
  name: string;
  mode: string;
  destinations: string[];
  members: string[];
  status: string;
}

export interface ConnectivityAnalysisSummary {
  id: string;
  tenant_id: string | null;
  device_id: string;
  mode: ConnectivityMode;
  device_b_id: string | null;
  status: ConnectivityStatus;
  anomaly_count: number;
  route_count: number;
  created_at: string;
  completed_at: string | null;
  error: string | null;
}

export interface ConnectivityAnalysisRead {
  id: string;
  tenant_id: string | null;
  device_id: string;
  mode: ConnectivityMode;
  device_b_id: string | null;
  status: ConnectivityStatus;

  // Dispositivo A
  routes: RouteEntry[] | null;
  bgp_peers: BgpPeer[] | null;
  ospf_neighbors: OspfNeighbor[] | null;
  sdwan_services: SdwanService[] | null;

  // Dispositivo B (apenas mode="pair")
  device_b_routes: RouteEntry[] | null;
  device_b_bgp_peers: BgpPeer[] | null;
  device_b_ospf_neighbors: OspfNeighbor[] | null;
  device_b_sdwan_services: SdwanService[] | null;

  anomalies: ConnectivityAnomalySeverity[] | null;
  ai_summary: string | null;
  ai_recommendations: string[] | null;
  error: string | null;
  created_at: string;
  completed_at: string | null;
}
